internal/stanza/converter.go

```diff
@@ -582,3 +571,60 @@ var sevTextMap = map[entry.Severity]string{
 	entry.Fatal3:  "Fatal3",
 	entry.Fatal4:  "Fatal4",
 }
+
+// pairSep is chosen to be an invalid byte for a utf-8 sequence
+// making it more unlikely to be hit
+var pairSep = []byte{0xfe}
+
+func getResourceID(resource map[string]string) uint64 {
+	var fnvHash = fnv.New64a()
+	var fnvHashOut = make([]byte, 0, 16)
+	var keySlice = make([]string, 0, len(resource))
+	var escapedSlice = make([]byte, 0, 64)
+
+	for k := range resource {
+		keySlice = append(keySlice, k)
+	}
+
+	// In order for this to be deterministic, we need to sort the map. Using range, like above,
+	// has no guarantee about order.
+	sort.Strings(keySlice)
+	for _, k := range keySlice {
+		escapedSlice = appendEscapedPairSeparator(escapedSlice[:0], k)
+		fnvHash.Write(escapedSlice)
+		fnvHash.Write(pairSep)
+
+		escapedSlice = appendEscapedPairSeparator(escapedSlice[:0], resource[k])
+		fnvHash.Write(escapedSlice)
+		fnvHash.Write(pairSep)
+	}
+
+	fnvHashOut = fnvHash.Sum(fnvHashOut)
+	return binary.BigEndian.Uint64(fnvHashOut)
+}
```

Comment:
Since the goal is optimization, we might consider accommodating some very common cases where even this function is unnecessary. Perhaps:
```go
if len(resource) == 0 {
  return []byte{ pairSep } // or some other individual byte
}

if len(resource) == 0 {
  // don't worry about sorting
  return ...
}
```

Reply:
Added in a check for this specific case. Let me know if I missed the point here, I just put this as the first check in the function.

Reply:
You got part of it. You may have seen my comment before I realized a typo and edited it. 

I meant to point out two special cases:
- `len(resource) == 0` is trivial because there's nothing to do, so just return immediately (you got this)
- `len(resource) == 1` is non-trivial, but could be treated as a special case that does not require any sortation or looping

I'm not sure how much benefit there is to either case, but I would not be surprised to find that a large portion of log events have either 0 or 1 resource attributes, so it may be worth optimizing both cases.

It also would be good to see before/after benchmarks so we know if the changes are actually helping.

Reply:
Yep, I missed the edit, haha. Alright, so heres the benchmarks:

Empty resource, before:
```
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/stanza
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkGetResourceIDEmptyResource-12    	24248468	        43.22 ns/op	      24 B/op	       1 allocs/op
```

Empty resource, after:
```
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/stanza
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkGetResourceIDEmptyResource-12    	514533082	         2.282 ns/op	       0 B/op	       0 allocs/op
PASS
```

Single resource, before:
```
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/stanza
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkGetResourceIDSingleResource-12    	 7344038	       160.9 ns/op	      40 B/op	       2 allocs/op
```

Single resource, after:
```
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/stanza
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkGetResourceIDSingleResource-12    	 9932624	       116.9 ns/op	      16 B/op	       1 allocs/op
```

Complex resource case, before:
```
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/stanza
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkGetResourceID-12    	 1000000	      1003 ns/op	     152 B/op	       2 allocs/op
```

Complex resource case, after:
```
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/stanza
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkGetResourceID-12    	 1000000	      1008 ns/op	     152 B/op	       2 allocs/op
```

So overall a good change, I'd say. Avoids an allocation in some cases, and doesn't really affect the more complex case really at all.

Reply:
That's great. Thanks for posting the benchmarks.